nvaccess/nvda
# Multiple dictionary dialogs can now be opened simultaneously if the dictionaries are of different types

Opened against beta branch since it is a follow-up of #12800.
Cc @seanbudd 

### Link to issue number:

Fix-up of #12800.
Described in https://github.com/nvaccess/nvda/pull/12800#issuecomment-911676751, paragraph "Test almost successful", "Test 2".

### Summary of the issue:

Since the merge of #12800, trying to re-open an already opened dialog allows to focus it rather than issuing an error as done previously. However, in the case of dictionaries, if you try to open the voice dictionary while the default is already open will lead to focus the default dictionary rather than opening the voice dictionary dialog;

### Description of how this pull request fixes the issue:

Since the check of an existing opened dialog is made on the class of the dialog, I have sub-classed `DictionaryDialog` for the three types of dictionaries to make them distinct one from another.

### Testing strategy:

Test 1:
* Tried to open the 3 dictionary dialogs simultaneously, either with the menu or with keyboard shortcut.

Test 2:
* Open the three dictionaries windows
* Updated them all (at least adding one entry for each)
* Close them all
* Re-open them all one by one to check that the content has been correctly updated.

### Known issues with pull request:

Maybe it would be interesting to forbid to instance an object of class `DictionaryDialog` directly, i.e. force subclassing.
I do not know however how it can be acheived and do not even know if it is pythonic.
Comments are welcome.

### Change log entries:
Not needed if merged along with #12800 in 2021.3.

### Code Review Checklist:

  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
